### PR TITLE
Add more config options to TokenPrinter

### PR DESCRIPTION
**Pull request by robinmaisch**
I'd like to propose two very small additions to the TokenPrinter that made it more helpful in my process of designing frontends. As it has been proposed to upload the TokenPrinter output for some sample files, this is the format that I would suggest for those output files.

1. Print code lines that contain no tokens if `PRINT_EMPTY_LINES` is active -- in case there are some lines where tokens are missing, or simply to get the complete context
2. Put an empty line after the last "token line" of a nonempty line if `SPACIOUS` is active -- this makes it easier on the eyes for me personally.

These flags are active by default in this PR, but the original behaviour can be restored by setting `PRINT_EMPTY_LINES` and `SPACIOUS` to false.

Sample:
```java
22      public final class TokenPrinter {
        |CLASS{

23          private static final Logger logger = LoggerFactory.getLogger(TokenPrinter.class);
            |VARDEF                                                                        |
                                                 |apply

24          // Representation:
25          private static final String BAR = "|";
            |VARDEF                             |

26          private static final String TAB = "\t";
            |VARDEF                              |

27          private static final String SPACE = " ";
            |VARDEF                               |
```
Before:
```java
22      public final class TokenPrinter {
        |CLASS{
23          private static final Logger logger = LoggerFactory.getLogger(TokenPrinter.class);
            |VARDEF                                                                        |
                                                 |apply
25          private static final String BAR = "|";
            |VARDEF                             |
26          private static final String TAB = "\t";
            |VARDEF                              |
27          private static final String SPACE = " ";
            |VARDEF                               |
```

### CODE REVIEW

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (tsaglam) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:115]: Avoid if conditions without curly braces.

**Review by tsaglam — CHANGES_REQUESTED** [review]
Just one small thing.

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (JanWittler) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:110]: I get that you tried to modify as little as possible with this PR and thus opted for this while loop. Before, the logic was to get the line for a token. However, now you are implicitly iterating over the lines and get the tokens for it, i.e. the logic changed to get all tokens for the current line. Thus, the code should be adapted to iterate over the lines which would also remove the need for a nested loop which is always difficult to read.

To keep the scope of this PR small, we can also make an issue and refactor the token printer in a subsequent PR. Probably @tsaglam should have the decided opinion on that 😄 .

Comment (tsaglam) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:110]: Both fixing it in here or making an issue is fine for me. To be fair, the token printer never was that readable to begin with. However, that's not a reason to leave it as is!

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (robinmaisch) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:120]: @jplag/maintainer 
I'm having a go at a refactoring, but I'm having trouble understanding why using the root folder as a replacement file name would be valid. Can someone explain?

Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:120]: Don't know neither but if you refactor it anyways, I would opt for doing it properly, i.e.
```Java
for (File file: files) {
  List<String> lines = getLinesForFile(file); //invented method but probably better than loading all lines upfront
  builder.append(System.lineSeparator());
  builder.append(file.getName());
  for (String line: lines) {
    / print lines and tokens
  }
}
```
and then you don't need the fallback anymore.

Comment (tsaglam) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:120]: Whenever your submission folder contains single files instead of folders, the root is the name of the file, and `Token.getFile()` is `""`.

This is a valid folder structure:
```
input/
|- submissionA.java
|- submissionB.java
|- submissionC.java
```

**Comment by robinmaisch** [comment]
This is my attempt at refactoring the TokenPrinter, and some small additions to go with it:

- Extracted everything concerning appending some output and managing the current position in the output to `PrinterOutputBuilder` (name open for debate)
- Extracted all data needed for the output of one line to `LineData`
   -> printTokens method is now much more concise

- Added method `stream` to `TokenList`, as that is used quite often in tests
- Added "filesAsSubmission" test to TokenPrinterTest

File: jplag.frontend-utils/src/main/java/de/jplag/TokenList.java
Comment (tsaglam) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenList.java:72]: I would rather change the return type of `allTokens()` to Collection rather than add a stream method. Collections can be directly streamed.

File: jplag.frontend.golang/src/test/java/de/jplag/golang/GoFrontendTest.java
Comment (tsaglam) [inline jplag.frontend.golang/src/test/java/de/jplag/golang/GoFrontendTest.java:160]: Why parallel? Is there really a performance issue here?

File: jplag.frontend.kotlin/src/test/java/de/jplag/kotlin/KotlinFrontendTest.java
Comment (tsaglam) [inline jplag.frontend.kotlin/src/test/java/de/jplag/kotlin/KotlinFrontendTest.java:156]: Also here.

File: jplag.frontend.rlang/src/test/java/de/jplag/rlang/RFrontendTest.java
Comment (tsaglam) [inline jplag.frontend.rlang/src/test/java/de/jplag/rlang/RFrontendTest.java:95]: Same here.

File: jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java
Comment (tsaglam) [inline jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java:164]: And here.

File: jplag/src/test/java/de/jplag/special/TokenPrinterTest.java
Comment (tsaglam) [inline jplag/src/test/java/de/jplag/special/TokenPrinterTest.java:82]: This is not ideal, as this class is not a test class in the traditional sense. If we want to unit test the token printer, we should actually do that in the module where the token printer resides. Also, no need to run JPlag to print tokens, you can just use the frontend itself and pass the tokens directly to the printer. This makes the unit test more fine-grained.

**Review by tsaglam — CHANGES_REQUESTED** [review]
I found a few things to discuss here.

**Comment by JanWittler** [comment]
Unfortunately, I can't annotate the lines that you didn't change but there are some minor things that should be fixed in the `TokenPrinter` while you are on it anyways.
- `printTokens(TokenList tokens, File directory, Collection<String> fileNames)` should just call `printTokens(tokens, directory, fileNames, Optional.empty())` instead of duplicating the string to file transformation
- the `root` / `directory` parameter in all `printTokens` methods should be renamed to `rootDirectory` consistently

**Review by JanWittler — COMMENTED** [review]
Code looks way better than before 👍🏼  (sorry Timur 😜 )

**Comment by tsaglam** [comment]
> Code looks way better than before 👍🏼 (sorry Timur 😜 )

I refer to this comment: https://github.com/jplag/JPlag/pull/594#discussion_r947773342 😄

File: jplag.frontend.golang/src/test/java/de/jplag/golang/GoFrontendTest.java
Comment (robinmaisch) [reply jplag.frontend.golang/src/test/java/de/jplag/golang/GoFrontendTest.java:160]: The only reason is that it was set to parallel before, and that the order does not matter here. But it is surely not necessary performance-wise.

File: jplag.frontend-utils/src/main/java/de/jplag/TokenList.java
Comment (robinmaisch) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenList.java:72]: Ah, sure, that's a good alternative! Coming right up.

File: jplag.frontend.golang/src/test/java/de/jplag/golang/GoFrontendTest.java
Comment (tsaglam) [reply jplag.frontend.golang/src/test/java/de/jplag/golang/GoFrontendTest.java:160]: Ah, I missed that, then it is a mistake that was already in there. Let's remove `parallel()` then.

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (robinmaisch) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:239]: Yes, but I thought it looked complicated with two `->` in  the same line.
```java
// TokenPrinter.java::96
lineDatas.forEach(lineData -> lineData.apply((String file, Integer lineNumber, String currentLine, List<Token> tokens) -> {
```
What do you think? Just to reiterate from the documentation, the purpose of the `apply` method is to avoid this:
```java
lineDatas.forEach(lineData -> {
	String file = lineData.file();
	int lineNumber = lineData.lineNumber();
	String currentLine = lineData.currentLine();
	List<Tokens> tokens = lineData.tokens();
	
	// Now, this function really begins.
	// ...
}
```
That's not wrong either, but it distracts from what's really happening.

Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:239]: As you access the values each only once (except for the current line), you could even avoid all these variables up-front and just use the accessors where needed. I would prefer that over the current solution as it is just simpler but that is just a subjective opinion. If you prefer your current solution, I am fine with that too.

File: jplag/src/test/java/de/jplag/special/TokenPrinterTest.java
Comment (robinmaisch) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java:82]: So, should TokenPrinterTest.java just be moved to the package de.jplag, or should a separate test class be created?

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (robinmaisch) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:239]: Ah, it doesn't look half as bad as I thought. I'll do just that.

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (robinmaisch) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:160]: Now, that makes this method much less intimidating. Great!
Weirdly enough, that makes the `file` parameter of `printTokens` unnecessary.

File: jplag/src/test/java/de/jplag/special/TokenPrinterTest.java
Comment (robinmaisch) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java:82]: Ah, got it. This new test class with the test I mentioned will make the frontend-utils dependent on at least one frontend (during testing only). But that is not an issue, I suppose.

Comment (tsaglam) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java:82]: Yes, that should lead to a cyclic dependency. From a software engineering standpoint, the best solution would be to make this test case frontend independent! Basically, create your own mock tokens with only the interfaces provided by the frontend-utils. The references file by the tokens could be anything, e.g., pseudo-code or just Java code. This has the advantage that the unit test does not depend on additional modules, thus keeping the dependencies clean. Moreover, this makes the unit test as fine-grained as possible, ensuring that it will only fail if the tested code changes and code that is not the subject under test, for example, a language frontend.

A sufficient test input could be two files, each containing one line. For those, you could just two tokens in the test case with trivial tokens.

Comment (robinmaisch) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java:82]: Ah yes, this is surely the best way to test the functionality of the TokenPrinter. I've come up with some fun test cases*.
The only question left for me is whether I should actively check for the correct output (e.g. by comparing to a hard-coded reference output file) or just accept the output as long as no exceptions occur. The correctness could still be checked "by eye" anytime.

*I realize that the original `printFilesAsSubmissions` test from above points to an issue separate from the TokenPrinter. Is it really intentional that the `file` attribute of tokens may be empty depending on the CLI arguments? This may change also, I think, so I'd rather not add a test that _requires_ the TokenPrinter to deal with the inconsistency.

Comment (tsaglam) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java]: You could just check if the printed string contains the lines of the file and the string representation of the token. This would be a relatively easy assertion to implement.

I think all tokens should point to a file, but maybe this is a weird edge case for single file submissions?

Comment (tsaglam) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java]: Also, as a comment: Don't put too much time into the token printer testing, as it is obviously not as important as other parts of our codebase. A simple test or a few simple tests should be enough for now.

**Comment by tsaglam** [comment]
@robinmaisch could you resolve all review comments that are already addressed?

File: jplag/src/test/java/de/jplag/special/TokenPrinterTest.java
Comment (robinmaisch) [reply jplag/src/test/java/de/jplag/special/TokenPrinterTest.java]: That's right. Done! I'll go over the SonarCloud issues and that should be it.

**Review by tsaglam — APPROVED** [review]

File: jplag.frontend-testutils/src/test/java/de/jplag/testutils/TokenUtils.java
Comment (SuyDesignz) [inline jplag.frontend-testutils/src/test/java/de/jplag/testutils/TokenUtils.java:32]: renaming was a good idea

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (SuyDesignz) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:33]: i'm not sure right now if moving the lines of code here makes sense. Was there a reason to move them?

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (SuyDesignz) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:71]: you have always used explicit typing except here. Best to keep consistent

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (SuyDesignz) [inline jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:194]: great idea!

File: jplag.frontend-utils/src/test/java/de/jplag/TestToken.java
Comment (SuyDesignz) [inline jplag.frontend-utils/src/test/java/de/jplag/TestToken.java:19]: good :+1:

File: jplag.frontend-utils/src/test/java/de/jplag/TokenPrinterTest.java
Comment (SuyDesignz) [inline jplag.frontend-utils/src/test/java/de/jplag/TokenPrinterTest.java:98]: good :+1:

File: jplag.frontend-utils/src/test/java/de/jplag/simple/TestTokenConstants.java
Comment (SuyDesignz) [inline jplag.frontend-utils/src/test/java/de/jplag/simple/TestTokenConstants.java:5]: I was a bit confused at the beginning because it contains only one parameter but in itself it is a good idea. :+1:

File: jplag.frontend.csharp-6/src/test/java/de/jplag/csharp/MinimalCSharpFrontendTest.java
Comment (SuyDesignz) [inline jplag.frontend.csharp-6/src/test/java/de/jplag/csharp/MinimalCSharpFrontendTest.java:55]: good 👍

File: jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java
Comment (SuyDesignz) [inline jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java:167]: good 👍

File: jplag.frontend.text/src/test/java/jplag/text/TextFrontendTest.java
Comment (SuyDesignz) [inline jplag.frontend.text/src/test/java/jplag/text/TextFrontendTest.java:49]: good 👍

File: jplag/src/test/java/de/jplag/TestBase.java
Comment (SuyDesignz) [inline jplag/src/test/java/de/jplag/TestBase.java:29]: nicer solution you have implemented here 👍

File: jplag/src/test/resources/de/jplag/samples/FilesAsSubmissions/Submission1.java
Comment (SuyDesignz) [inline jplag/src/test/resources/de/jplag/samples/FilesAsSubmissions/Submission1.java:11]: Test data looks good  👍

**Review by SuyDesignz — APPROVED** [review]
I could not find any real problems or inconsistencies in the code. I have left a few comments but in itself it all fits very well 👍

File: jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java
Comment (robinmaisch) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:33]: That must have been the Spotless plugin. You are right, that looks weird.

Comment (robinmaisch) [reply jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java:33]: I didn't set them public on purpose, so it's good that this came up.

**Comment by tsaglam** [comment]
@robinmaisch, either let me know if this is ready to merge to resolve all comments that were addressed so I know that I can merge.

**Comment by robinmaisch** [comment]
@tsaglam Yes, it is, please go ahead. I'll merge with Rust right after and update the demo files.

**Pull request merged**
